airbyte-ci/connectors/pipelines/tests/test_helpers/test_utils.py

```diff
@@ -241,3 +242,12 @@ async def test_export_container_to_tarball_failure(mocker, tmp_path):
         str(tmp_path / f"my_connector_my_git_revision_{platform.replace('/', '_')}.tar"),
         forced_compression=dagger.ImageLayerCompression.Gzip,
     )
+
+
+@pytest.mark.anyio
+async def test_get_repo_dir(mocker, dagger_client):
+    test_context = PipelineContext(pipeline_name="test", is_local=True, git_branch="test", git_revision="test", report_output_prefix="test")
+    test_context.dagger_client = dagger_client
+    filtered_directory = await test_context.get_repo_dir("airbyte-ci/connectors/")
+    unfiltered_directory = await dagger_client.host().directory("airbyte-ci/connectors/")
+    raise Exception("" + str((await filtered_directory.entries())) + "--" + str((await unfiltered_directory.entries())))
```

Comment:
@stephane-airbyte you told me in DM that something is going wrong with this test? Could you please share more details? Raising here will fail the test.
Is the problem that filtered and unfiltered entries are the same?

Reply:
yes, that's exactly my problem. And that's without my changes to the filter logic. So either I'm missing something (likely) or the filter is useless

Reply:
I think I have some better understanding of what was going on.
The filter starts on the current working dir, and it's always being set to the directory from which I was executing poetry.
In pycharm I'm able to change that by modifying the working directory in the running configuration. But basically, I can't get this test to pass from the command line.
That being said, when running it from pycharm, the result is that `dagger_client.host().directory()` doesn't support the `**` construct. So I ended up implemented what you suggested, with a cache, so we only call  glob 12 (!) times per instance of PipelineContext

Reply:
That being said, I'm not sure that the current dependency on the working directory is a desired behavior. I don't have enough understanding of the code to be definitive, but I wonder if we shouldn't use the `root_dir` parameter to `glob`.
Finally, I think we could improve the performance further by ourselves filtering the values out of glob, so we only walk the filesystem once.

Reply:
I couldn't resist but try to avoid doing so many calls to glob. See the `faster` function. I checked the returned set versus the original one, and they're the same. I haven't measured the speed difference, but I'd expect it to be another 12x. Now, the bottleneck for format seems to be the time to get a dagger_client

Reply:
> That being said, I'm not sure that the current dependency on the working directory is a desired behavior.

I can't say it desired but it's a fact - a lot of the logic in this codebase depends on the repo structure. This is why we call[ this function very early](https://github.com/airbytehq/airbyte/blob/main/airbyte-ci/connectors/pipelines/pipelines/cli/ensure_repo_root.py#L55) in the execution path . It'd be interesting to call it in your tests.

Reply:
I actually did call this function first thing in my test, and still the working dir was somehow set to the directory from which the tests were started...

Reply:
I'm going to just @lru_cache the previous logic. My attempt depends on python 3.11+, which we don't seem to have everywhere :(